### PR TITLE
in upath_io_manager, set no path metadata when obj is None

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

Usually, when an op or asset returns `None`, it means that the user didn't intend to use an IO manager, but ended up with one because they didn't set `-> None` as their return type annotation. In these cases, seeing a "path" metadata value can be very confusing, because usually the body of their op is actually writing data to a different location.

We've discussed solutions at length that involved bypassing the IO manager, but ultimately backed off from them due to backwards compatibility concerns.

This PR proposes a smaller step: omitting the metadata when the outputted in object is `None`. This avoids the confusion in the common case. There's a tradeoff here: this comes at the cost of consistency. And it means that, in the rare cases where people actually intend to return `None` and have it written to a file, they won't see the metadata for that file.

### CODE REVIEW

**Review by jamiedemaria — APPROVED** [review]
good w me!

**Review by schrockn — CHANGES_REQUESTED** [review]
Seems like if we do this we should also not write out a file, for consistency.

We would need to change logic on the read side to not a non-existent file, but it seems like we could do that with some level of metadata tracking.

What do you think?

**Comment by sryza** [comment]
@jamiedemaria @alangenfeld and I went deep down that path (I believe both skipping the IO manager entirely and special-casing the default IO manager, as you suggested) a couple months ago. It seemed very promising at first, but there were some gnarly backcompat cases we ran into, and we ultimately backed off.  So this PR is basically a consolation prize.

@jamiedemaria - do you happen to remember why we didn't go with the approach @schrockn mentioned above?

**Comment by sryza** [comment]
It might have been a cross code location compatibility issue. E.g.
- Asset 1 in code location A is upstream of asset 2 in code location B
- Code location A upgrades to the latest version
- Code location B is still on an older version
- When we materialize asset 1, we do the new thing of writing metadata instead of a file
- When we try to materialize asset 2, we try to load asset 1, but it fails because there's no file written

**Comment by jamiedemaria** [comment]
This pretty long comment I left on a PR that tried to skip the store/load of `None`s explains why we couldn't do the metadata tracking thing https://github.com/dagster-io/dagster/pull/18820#issuecomment-1930033192

tldr:
Adding a marker to metadata would require that we add a new fetch from the DB to load the `AssetMaterialization` event at load time. We decided it was too critical a path to add additional DB fetches.

Adding a marker to the tags (to piggy back off of the fetches already being done by the versioning system) wont work because the versioning system doesn't fetch anything if there are more that 100 upstream partitions being mapped to the downstream. But the I/O manager would still need to fetch those tags which could overload the DB

**Comment by schrockn** [comment]
> Adding a marker to metadata would require that we add a new fetch from the DB to load the AssetMaterialization event at load time.

I think loading the upstream asset materialization for every input is totally reasonable (and I would bet happens most of the time anyways.)

Is the actual concern cases where we are loading an event for every upstream asset _partition_?

**Comment by jamiedemaria** [comment]
yeah partitions is where it gets gnarly. i think there were some concerns about fetching the event even for non-partitioned assets, but i don't remember specifics and can't find anything written, so i could be misremembering

**Comment by schrockn** [comment]
Got it. Ok let's move forward with this. I still think we need to address this more fundamentally but this is a good mitigation.

@jamiedemaria could you assemble all of this context and create a linear issue?

**Review by schrockn — APPROVED** [review]

**Comment by jamiedemaria** [comment]
made an issue here https://linear.app/dagster-labs/issue/FOU-179/investigate-foundational-work-required-to-allow-io-managers-to-skip i'll look through all the docs i have on the subject and update the issue/linked notion page to function as a source of truth
